packages/core/src/agent/MessageSender.ts

```diff
@@ -408,13 +419,59 @@ export class MessageSender {
     outboundPackage.endpoint = service.serviceEndpoint
     outboundPackage.connectionId = connection?.id
-    for (const transport of this.outboundTransports) {
-      const protocolScheme = getProtocolScheme(service.serviceEndpoint)
-      if (!protocolScheme) {
-        this.logger.warn('Service does not have valid protocolScheme.')
-      } else if (transport.supportedSchemes.includes(protocolScheme)) {
-        await transport.sendMessage(outboundPackage)
-        return
+
+    // Outbound transport is present
+    if (this.outboundTransports.length !== 0) {
+      for (const transport of this.outboundTransports) {
+        const protocolScheme = getProtocolScheme(service.serviceEndpoint)
+        if (!protocolScheme) {
+          this.logger.warn('Service does not have a protocol scheme.')
+        } else if (transport.supportedSchemes.includes(protocolScheme)) {
+          await transport.sendMessage(outboundPackage)
+          return
+        }
+      }
+      // No outbound transport: Try retrieving session-based connection instead
+    } else if (sessionId || sessionIdFromInbound) {
+      const session = sessionIdFromInbound
+        ? this.transportService.findSessionById(sessionIdFromInbound)
+        : sessionId
+        ? this.transportService.findSessionById(sessionId)
+        : undefined
+
+      if (session?.inboundMessage?.hasReturnRouting(message.threadId)) {
+        this.logger.debug(
+          `Found session with return routing for message '${message.id}' (connection '${connection?.id}'`
+        )
+        try {
+          await this.sendMessageToService(outboundMessageContext)
+          this.emitMessageSentEvent(outboundMessageContext, OutboundMessageSendStatus.SentToSession)
+          return
+        } catch (error) {
+          this.logger.debug(`Sending an outbound message via session failed with error: ${error.message}.`, error)
+          throw new MessageSendingError(
+            `Unable to send message to service: ${service.serviceEndpoint} ${message.threadId} ${JSON.stringify(
+              session
+            )}`,
+            {
+              outboundMessageContext,
+            }
+          )
+        }
+      } else if (outboundMessageContext.serviceParams?.service.serviceEndpoint && session) {
+        try {
+          await this.sendMessageToSession(agentContext, session, message)
+          this.emitMessageSentEvent(outboundMessageContext, OutboundMessageSendStatus.SentToSession)
+          return
+        } catch (error) {
+          this.logger.error(`Sending an outbound message via session failed with error: `, error)
+          throw new MessageSendingError(
+            `Unable to send message to service: ${outboundMessageContext.serviceParams?.service.serviceEndpoint} with threadId ${message.threadId} `,
+            {
+              outboundMessageContext,
+            }
+          )
+        }
```

Comment:
We shouldn't send it over the session if a serviceEndpoint was present. It's only allowed to use the session if return routing is enabled, and this is something that should be fixed from the other side (so in this case the presentation MUST have return route).

We should add it to the presentation handlers (for both v1 and v2):
https://github.com/morrieinmaas/aries-framework-javascript/blob/feat/outbound-message-send-via-session/packages/core/src/modules/proofs/protocol/v2/handlers/V2PresentationHandler.ts#L56-L62

If you see the difference with the proofs Api, we explicitly pass that returnRoute should be used: https://github.com/morrieinmaas/aries-framework-javascript/blob/feat/outbound-message-send-via-session/packages/core/src/modules/proofs/ProofsApi.ts#L498